security/pkg/pki/ca/ca.go

```diff
@@ -317,9 +329,244 @@ func (ca *IstioCA) SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl tim
+	}
+
+	status := ca.checkAndRotateRootCertForSigningCertCitadel(config, caSecret, scrtErr)
+	if status == UpgradeSuccess {
+		config.Metrics.RootUpgradeSuccess.Increment()
+	}
+	if status == UpgradeFailure {
+		config.Metrics.RootUpgradeErrors.Increment()
+	}
+	return
+}
+
+// loadCASecretWithRetry lets the self-signed Citadel read CA secret with retries until
+// timeout.
+func (cs *IstioCA) loadCASecretWithRetry(config *SelfSignedCARootCertRotationConfig) (*v1.Secret, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), config.CheckInterval/4)
+	defer cancel()
+
+	caSecret, scrtErr := config.Client.Secrets(config.CaStorageNamespace).Get(CASecret, metav1.GetOptions{})
+	if scrtErr != nil && !errors.IsNotFound(scrtErr) {
+		log.Errorf("Self-signed Citadel failed to read secret that holds CA certificate: %s. "+
+			"Wait until secret %s:%s can be loaded", scrtErr.Error(), config.CaStorageNamespace, CASecret)
+		ticker := time.NewTicker(config.RetryInterval)
+		for scrtErr != nil {
+			select {
+			case <-ticker.C:
+				if caSecret, scrtErr = config.Client.Secrets(config.CaStorageNamespace).Get(CASecret, metav1.GetOptions{}); scrtErr == nil {
+					break
+				}
+			case <-ctx.Done():
+				log.Errorf("Self-signed Citadel failed to load CA secret %s:%s until timeout.",
+					config.CaStorageNamespace, CASecret)
+				break
+			}
+		}
+	}
+	return caSecret, scrtErr
+}
+
+func (ca *IstioCA) checkAndRotateRootCertForReadOnlyCitadel(
+	config *SelfSignedCARootCertRotationConfig, caSecret *v1.Secret, scrtErr error) RootUpgradeStatus {
+	if scrtErr != nil {
+		log.Errorf("Citadel runs in self-signed root cert read only mode but"+
+			" fail to load secret %s:%s (error: %s), skip cert rotation job",
+			config.CaStorageNamespace, CASecret, scrtErr.Error())
+		return UpgradeSkip
+	}
+
+	if caSecret == nil {
+		log.Info("Root cert does not exist, skip root cert rotate for " +
+			"self-signed root cert read-only Citadel.")
+		return UpgradeSkip
+	}
+
+	rootCertificate := ca.GetCAKeyCertBundle().GetRootCertPem()
+	if !bytes.Equal(rootCertificate, caSecret.Data[RootCertID]) {
+		// If the CA secret holds a different root cert than the root cert stored in
+		// KeyCertBundle, this indicates that the local stored root cert is not
+		// up-to-date. Update root cert and key in KeyCertBundle and config map.
+		log.Infof("Load signing key and cert from existing secret %s:%s", caSecret.Namespace, caSecret.Name)
+		rootCerts, err := appendRootCerts(caSecret.Data[caCertID], config.RootCertFile)
+		if err != nil {
+			log.Errorf("failed to append root certificates (%v)", err)
+			return UpgradeFailure
+		}
+		keyCertBundle, err := util.NewVerifiedKeyCertBundleFromPem(caSecret.Data[caCertID],
+			caSecret.Data[caPrivateKeyID], nil, rootCerts)
+		if err != nil {
+			log.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			return UpgradeFailure
+		}
+
+		if err = updateCACertInConfigmapWithRetry(config, keyCertBundle.GetRootCertPem()); err != nil {
```

Comment:
why we think it's not okay to have read only mode citadel write self signed root secret, but here we let read only citadel touching the config map of the root cert as well?

this seems introduce race condition again?

Reply:
I have changed the read only Citadel to not touch config map. Only the Citadel that generates new CA root cert will update config map.